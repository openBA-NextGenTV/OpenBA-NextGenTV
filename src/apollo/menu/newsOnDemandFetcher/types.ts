export type FeedItem = {
  id: string;
  title: string;
  duration: number;
  videoUrl: string;
  thumbnailUrl: string;
  category: string;
};
