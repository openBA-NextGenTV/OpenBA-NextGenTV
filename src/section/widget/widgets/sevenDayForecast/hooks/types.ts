export type ResultData = {
  city: string;
  state: string;
  headerTemp: number;
  headerImage: number;
  headerTitle: string;
  days: Day[];
};

export type Day = {
  dateTime: number;
  image: number;
  title: string;
  hiTemp: number;
  lowTemp: number;
  precipitation: number;
  wind: number;
  windDirection: string;
  humidity: number;
  ultraviolet: string;
};
