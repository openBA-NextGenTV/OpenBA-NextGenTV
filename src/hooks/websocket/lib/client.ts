/*
 * Copyright 2021 OpenBA-NextGenTV Contributors (https://OpenBA-NextGenTV.tech)
 * Copyright 2021 Sinclair Broadcast Group, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { EventEmitter } from 'events';
import {
  format,
  InvalidRequest,
  JsonRpcPayload,
  JsonRpcPayloadError,
  JsonRpcPayloadNotification,
  JsonRpcPayloadRequest,
  JsonRpcPayloadResponse,
  MethodNotFound,
  parse,
} from 'json-rpc-protocol';

import { ClientOptions, ReconnectOptions } from './client.types';
import { InternalError, MaxReconnectAttemptsExceedError } from './errors';

const RESERVED_EVENTS = [
  'error',
  'open',
  'close',
  'message',
  'unexpected-response',
  'upgrade',
  'reopen',
  'protocol:error',
  'protocol:response',
  'protocol:request',
];

export class Client extends EventEmitter {
  private client?: WebSocket;

  private options: ClientOptions;
  private reconnect: ReconnectOptions;
  private requestId = 0;

  constructor(public endpoint: string, options?: ClientOptions) {
    super();

    this.options = Object.assign(
      {
        callTimeout: 5000,
        reconnectTimeout: 5000,
        reconnectAttempts: Infinity,
        waitConnectionTimeout: 5000,
      },
      options,
    );

    this.reconnect = {
      attempted: 0,
      timeoutId: null,
    };

    this.open();
  }

  open() {
    if (this.client) {
      this.emit('error', new InternalError("Client exists, can't open a new websocket connection"));
      return;
    }

    this.client = new WebSocket(this.endpoint);

    this.client.addEventListener('error', error => this.onError(error));
    this.client.addEventListener('open', () => this.onOpen());
    this.client.addEventListener('close', ({ code, reason }) => this.onClose(code, reason));
    this.client.addEventListener('message', message => this.onMessage(message));
  }

  close(code?: number, reason?: string) {
    this.cleanupReconnect();

    if (!this.client) {
      this.emit('error', new InternalError("Client missing, can't close websocket connection"));
      return;
    }

    this.client.close(code, reason);
  }

  isOpen() {
    return !!(this.client && this.client.readyState === this.client.OPEN);
  }

  call(method: string, params?: any) {
    const id = ++this.requestId;

    let message: string;

    try {
      message = format.request(id, method, params);
    } catch (error) {
      this.emit('error', error);
      return Promise.reject(error);
    }

    return this.waitConnection().then(
      () =>
        new Promise((resolve, reject) => {
          this.client?.send(message);

          const timeout = window.setTimeout(() => {
            reject(new InternalError(`Call on method '${method}' timed out`));
          }, this.options.callTimeout);

          const protocolResponseHandler = (receivedId: number, result: any) => {
            if (receivedId === id) {
              cleanupCallEvents();
              resolve(result);
            }
          };

          const protocolErrorHandler = (receivedId: number, protocolError: any) => {
            if (receivedId === id) {
              cleanupCallEvents();
              reject(protocolError);
            }
          };

          const closeHandler = () => {
            const closeError = new InternalError(`Call on method ${method} failed, broken connection`);

            cleanupCallEvents();
            reject(closeError);
          };

          const cleanupCallEvents = () => {
            clearTimeout(timeout);

            this.removeListener('protocol:response', protocolResponseHandler);
            this.removeListener('protocol:error', protocolErrorHandler);

            this.removeListener('close', closeHandler);
          };

          this.on('protocol:response', protocolResponseHandler);
          this.on('protocol:error', protocolErrorHandler);

          this.on('close', closeHandler);
        }),
    );
  }

  private waitConnection(): Promise<void> {
    if (this.isOpen()) {
      return Promise.resolve();
    }

    return new Promise((resolve, reject) => {
      const handler = () => {
        clearTimeout(timeoutId);
        this.removeListener('open', handler);

        this.isOpen() ? resolve() : reject(new InternalError(`Client gone, can't request`));
      };

      const timeoutId = window.setTimeout(handler, this.options.waitConnectionTimeout);
      this.once('open', handler);
    });
  }

  /* Client events */
  private onError(error: Event) {
    this.emit('error', error);
  }

  private onOpen() {
    this.cleanupReconnect();

    this.emit('open');
    // this.emit('org.atsc.notify', { alertingList: [] });
  }

  private onClose(code?: number, reason?: string) {
    delete this.client;
    this.emit('close', code, reason);

    console.log('WS connection was closed by server.');

    this.performReconnect();
  }

  private onMessage(message: MessageEvent) {
    let parsedMessage;

    try {
      parsedMessage = parse(message.data);
    } catch (error) {
      this.emit('error', error);
      return;
    }

    this.handleMessage(parsedMessage as JsonRpcPayload);
  }

  /* Reconnect */

  private performReconnect() {
    if (this.reconnect.attempted >= this.options?.reconnectAttempts) {
      this.emit('error', new MaxReconnectAttemptsExceedError('Too many failed connection attempts'));
      return;
    }

    this.reconnect.timeoutId = window.setTimeout(() => {
      this.reconnect.attempted += 1;

      this.emit('reopen', this.options.reconnectAttempts, this.reconnect.attempted);

      this.open();
    }, this.options.reconnectTimeout);
  }

  /* Protocol */

  private cleanupReconnect() {
    if (this.reconnect.timeoutId) {
      clearTimeout(this.reconnect.timeoutId);
    }

    this.reconnect = {
      attempted: 0,
      timeoutId: null,
    };
  }

  private handleMessage(payload: JsonRpcPayload) {
    if ('method' in payload && RESERVED_EVENTS.includes(payload.method)) {
      this.emit('error', new MethodNotFound(`Reserved method call: ${payload.method}`));
      return;
    }

    if ('method' in payload && payload.method === 'org.atsc.notify') {
      const { method, params } = payload as JsonRpcPayloadNotification;
      this.emit(method, params);
      return;
    }

    const { type } = payload;

    switch (type) {
      case 'error': {
        const { id, error } = payload as JsonRpcPayloadError;
        this.emit('protocol:error', id, error);
        break;
      }

      case 'response': {
        const { id, result } = payload as JsonRpcPayloadResponse;
        this.emit('protocol:response', id, result);
        break;
      }

      case 'request': {
        const { method, id, params } = payload as JsonRpcPayloadRequest;
        this.emit('protocol:request', method, id, params);
        break;
      }

      case 'notification': {
        const { method, params } = payload as JsonRpcPayloadNotification;
        this.emit(method, params);
        break;
      }

      default:
        this.emit('error', new InvalidRequest(`Unknown protocol message type: ${type}`));
        break;
    }
  }
}
