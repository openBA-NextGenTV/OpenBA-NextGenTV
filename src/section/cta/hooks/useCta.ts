/*
 * Copyright 2021 OpenBA-NextGenTV Contributors (https://OpenBA-NextGenTV.tech)
 * Copyright 2021 Sinclair Broadcast Group, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useEffect, useState } from 'react';

import { useAlertsModel, useDeviceInfoModel, useFlashChannelModel } from '../../../state';

let ctaIsVisibleTimeoutId: number | undefined;

export const useCta = () => {
  const [isVisible, setIsVisible] = useState(true);
  const { alert } = useAlertsModel();
  const { flashChannel } = useFlashChannelModel();
  const { station } = useDeviceInfoModel();
  const logoUrl = station ? `stations/${station}/logo.png` : null;

  useEffect(() => {
    /*
    1. If we already show CTA button - don't show it again.
    2. If alert is present in AlertBar - don't show CTA button
    3. Let wait 5 seconds and hide CTA button
     */
    if (!isVisible) {
      return;
    }

    if (alert || flashChannel) {
      setIsVisible(false);
    } else {
      if (ctaIsVisibleTimeoutId) {
        clearTimeout(ctaIsVisibleTimeoutId);
      }

      ctaIsVisibleTimeoutId = window.setTimeout(() => {
        setIsVisible(false);
      }, 5000);
    }
  }, [alert, flashChannel, isVisible]);

  return {
    isVisible,
    logoUrl,
  };
};
