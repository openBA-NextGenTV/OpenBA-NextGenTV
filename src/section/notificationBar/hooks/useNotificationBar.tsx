import { useEffect, useState } from 'react';

import {
  useAlertsModel,
  useAlertsOperations,
  useDeviceInfoModel,
  useFlashChannelModel,
  useFlashChannelOperations,
  useMenusModel,
  useMenusOperations,
  useWidgetModel,
} from '../../../state';
import { NotificationProps } from '../Notification';
import { getAlertProps, getFlashChannelEndProps, getFlashChannelProps } from '../notification.util';

export const useNotificationBar = () => {
  const [notificationProps, setNotificationProps] = useState<null | NotificationProps>(null);
  const [isHiddenEndFlashChannelBar, setIsHiddenEndFlashChannelBar] = useState(false);

  const { alert } = useAlertsModel();
  const { hideAlert, showMore } = useAlertsOperations();
  const { hideFlashChannel, gotoChannel } = useFlashChannelOperations();
  const { flashChannel } = useFlashChannelModel();
  const { station } = useDeviceInfoModel();

  const { isVisible } = useMenusModel();
  const { widget } = useWidgetModel();

  const { disableMenu, enableMenu } = useMenusOperations();

  const hideEndFlashChannel = () => {
    setIsHiddenEndFlashChannelBar(true);
    setNotificationProps(null);
  };

  useEffect(() => {
    if (flashChannel?.isLive) {
      const flashChannelProps = getFlashChannelProps(flashChannel, hideFlashChannel, gotoChannel);
      setIsHiddenEndFlashChannelBar(false);
      setNotificationProps(flashChannelProps);
      disableMenu();
      return;
    }

    if (flashChannel && flashChannel?.expireTime && !isHiddenEndFlashChannelBar) {
      const endFlashChannelProps = getFlashChannelEndProps(hideEndFlashChannel, station, flashChannel.title);
      setNotificationProps(endFlashChannelProps);
      disableMenu();
      return;
    }

    if (alert) {
      const alertProps = getAlertProps(alert, hideAlert, showMore);
      setNotificationProps(alertProps);
      disableMenu();
      return;
    }

    enableMenu();
    setNotificationProps(null);
  }, [
    isVisible,
    station,
    widget,
    alert,
    flashChannel,
    isHiddenEndFlashChannelBar,
    gotoChannel,
    hideAlert,
    hideFlashChannel,
    showMore,
    disableMenu,
    enableMenu,
  ]);

  return {
    notificationProps,
  };
};
