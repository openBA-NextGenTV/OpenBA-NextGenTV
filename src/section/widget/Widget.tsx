/*
 * Copyright 2021 OpenBA-NextGenTV Contributors (https://OpenBA-NextGenTV.tech)
 * Copyright 2021 Sinclair Broadcast Group, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FC, useEffect } from 'react';
import ReactGA from 'react-ga';

import { GetWidgetQuery, Widget as WidgetModel } from '../../apollo/generated/graphql';
import { useDeviceInfoModel, useWidgetModel } from '../../state';
import {
  Alert,
  AudioPlayerWrapper,
  FileList,
  LatestWeathercast,
  Priority,
  PrivacyPolicy,
  RMPPlaceholderWrapper,
  SevenDayForecast,
  SystemInfo,
  VideoPlayerWrapper,
  Zip,
} from './widgets';
import { Radio } from './widgets/otaRadio';

export type WidgetComponentProps = {
  widget: WidgetModel;
};

export const Widget = () => {
  const { widget } = useWidgetModel();

  useTrackPage(widget);

  if (!widget) {
    return <RMPPlaceholderWrapper />;
  }

  let WidgetComponent: FC<WidgetComponentProps> | null = null;
  let needToShowRMPPlaceHolder = false;

  switch (widget.type) {
    case 'alert':
      WidgetComponent = Alert;
      break;
    case 'audioPlayer':
      WidgetComponent = AudioPlayerWrapper;
      break;
    case 'videoPlayer':
      WidgetComponent = VideoPlayerWrapper;
      break;
    case 'zip':
      WidgetComponent = Zip;
      needToShowRMPPlaceHolder = true;
      break;
    case 'priority':
      WidgetComponent = Priority;
      needToShowRMPPlaceHolder = true;
      break;
    case 'privacyPolicy':
      WidgetComponent = PrivacyPolicy;
      break;
    case 'systemInfo':
      WidgetComponent = SystemInfo;
      needToShowRMPPlaceHolder = true;
      break;
    case 'sevenDayForecast':
      WidgetComponent = SevenDayForecast;
      break;
    case 'latestWeathercast':
      WidgetComponent = LatestWeathercast;
      break;
    case 'fileList':
      WidgetComponent = FileList;
      break;
    case 'otaRadio':
      WidgetComponent = Radio;
      break;
  }

  if (!WidgetComponent) {
    return null;
  }

  return (
    <>
      <WidgetComponent widget={widget} />

      {needToShowRMPPlaceHolder && <RMPPlaceholderWrapper />}
    </>
  );
};

const useTrackPage = (widget: GetWidgetQuery['widget']) => {
  const { serviceId } = useDeviceInfoModel();

  useEffect(() => {
    if (widget && widget.metric) {
      const { page, title } = widget.metric;

      ReactGA.pageview(`${serviceId}/${page}`, undefined, title || undefined);
    }
  }, [widget, serviceId]);
};
