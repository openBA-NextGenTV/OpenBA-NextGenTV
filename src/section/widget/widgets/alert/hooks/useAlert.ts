/*
 * Copyright 2021 OpenBA-NextGenTV Contributors (https://OpenBA-NextGenTV.tech)
 * Copyright 2021 Sinclair Broadcast Group, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import objectHash from 'object-hash';
import { useCallback, useEffect, useState } from 'react';

import { Alert, Page } from '../../../../../apollo/generated/graphql';
import {
  Command,
  CommandListener,
  registerView,
  resumeRmpPlayback,
  stopRmpPlayback,
  unregisterView,
} from '../../../../../hooks';
import { useMenusOperations, useWidgetOperations } from '../../../../../state';
import { isMobile } from '../../../../../utils';
import { BackButtonRenderer, TitleRenderer } from '../sideMenu/renderer';

export type PageItemModel = {
  title?: string;
  selected?: boolean;
  highlighted?: boolean;
  payload?: Page;
  renderer: any;
};

export const useAlert = (alert: Alert) => {
  const { closeWidget } = useWidgetOperations();

  const [pages, setPages] = useState(getPages(alert));
  const [alertHash, setAlertHash] = useState(objectHash(alert));
  const [pageActive, setPageActive] = useState(false);

  useEffect(() => {
    if (alertHash !== objectHash(alert)) {
      setAlertHash(objectHash(alert));
      setPages(getPages(alert));
    }
  }, [alert, alertHash]);

  useEffect(() => {
    stopRmpPlayback();
    return () => {
      resumeRmpPlayback();
    };
  }, []);

  const selectHighlightedPage = useCallback(
    () => setPages(pages.map(value => ({ ...value, selected: value.highlighted }))),
    [setPages, pages],
  );

  const highlightPage = useCallback(
    (direction: 'nextPage' | 'previousPage') => {
      const selectedIndex = pages.findIndex(page => page.highlighted);
      const nextIndex =
        direction === 'nextPage'
          ? selectedIndex === pages.length - 1
            ? 0
            : selectedIndex + 1
          : selectedIndex === 0
          ? pages.length - 1
          : selectedIndex - 1;

      pages.forEach((page, index) => (page.highlighted = index === nextIndex));

      if (pages[nextIndex].title !== 'BackToAlerts') {
        pages.forEach(page => (page.selected = false));
        pages[nextIndex].selected = true;
      }

      setPages([...pages]);
    },
    [pages, setPages],
  );

  const pageClickHandler = useCallback(
    (page: PageItemModel) => {
      if (page.title === 'BackToAlerts') {
        closeWidget();
      } else {
        pages.forEach(value => (value.highlighted = false));
        page.highlighted = true;
        selectHighlightedPage();
        setPageActive(false);
      }
    },
    [closeWidget, pages, selectHighlightedPage],
  );

  const commandListener = useCallback(
    (command: Command) => {
      switch (command) {
        case 'ArrowUp':
          highlightPage('previousPage');
          break;
        case 'ArrowDown':
          highlightPage('nextPage');
          break;
        case 'ArrowRight':
          setPageActive(true);
          break;
        case 'Enter':
          if (pages[0]['highlighted']) {
            closeWidget();
          } else {
            setPageActive(true);
          }
          break;
        case 'ArrowLeft':
        case 'Backspace': {
          if (!pageActive) {
            closeWidget();
          }
          setPageActive(false);
          break;
        }
      }

      return false;
    },
    [pages, selectHighlightedPage, highlightPage, closeWidget, pageClickHandler],
  );

  useRegisterView(commandListener);

  useSelectAlertMenuAfterUnmount(alert);

  return {
    selectedPage: pages.find(value => value.selected),
    pages,
    pageClickHandler,
    pageActive,
    setPageActive,
  };
};

const getPages = (alert: Alert) => {
  const result: PageItemModel[] = [
    { title: 'BackToAlerts', renderer: BackButtonRenderer },
    ...alert.pages.map(page => ({ title: page.title, payload: page, renderer: TitleRenderer })),
  ];

  // select the first page by default
  if (result.length >= 2) {
    result[1].selected = true;
    result[1].highlighted = true;
  }

  return result;
};

const useRegisterView = (commandListener: CommandListener) => {
  useEffect(() => {
    const viewId = 'AlertId';

    registerView({ viewId: viewId, listener: commandListener });

    return () => unregisterView({ viewId: viewId });
  }, [commandListener]);
};

const useSelectAlertMenuAfterUnmount = (alert: Alert) => {
  const { showMenu, selectMenu } = useMenusOperations();

  useEffect(() => {
    return () => {
      showMenu();

      if (isMobile()) {
        selectMenu('alerts');
      }
    };
  }, [showMenu, selectMenu, alert]);
};
