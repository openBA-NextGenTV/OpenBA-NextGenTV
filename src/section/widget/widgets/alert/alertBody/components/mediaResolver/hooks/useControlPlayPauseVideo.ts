/*
 * Copyright 2021 OpenBA-NextGenTV Contributors (https://OpenBA-NextGenTV.tech)
 * Copyright 2021 Sinclair Broadcast Group, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useCallback, useEffect, useRef } from 'react';

import { Command, registerView, unregisterView } from '../../../../../../../../hooks';

export const useControlPlayPauseVideo = (pageActive: boolean, setPageActive: (a: boolean) => void) => {
  const videoRef = useRef<HTMLVideoElement | null>(null);

  const playPause = () => {
    if (videoRef.current?.paused) {
      videoRef.current?.play();
    } else {
      videoRef.current?.pause();
    }
  };
  const commandListener = useCallback(
    (command: Command) => {
      switch (command) {
        case 'Enter': {
          if (videoRef.current && pageActive) {
            playPause();
          }
          break;
        }
        case 'ArrowLeft': {
          setPageActive(false);
        }
      }

      return false;
    },
    [pageActive, setPageActive],
  );

  useEffect(() => {
    if (pageActive) {
      registerView({ viewId: 'MediaResolver', listener: commandListener });
    } else {
      unregisterView({ viewId: 'MediaResolver' });
    }
    return () => {
      unregisterView({ viewId: 'MediaResolver' });
    };
  }, [pageActive, commandListener]);

  const mediaClickHandler = () => {
    if (videoRef.current && pageActive) {
      playPause();
    } else {
      setPageActive(true);
    }
  };

  return { videoRef, mediaClickHandler };
};
