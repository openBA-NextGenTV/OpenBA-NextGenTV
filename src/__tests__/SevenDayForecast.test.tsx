import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { SevenDayForecast } from '../section/widget/widgets';
import { ResultData } from '../section/widget/widgets/sevenDayForecast/hooks/types';
import { WeatherDay } from '../section/widget/widgets/sevenDayForecast/weatherDay';
import { Property } from '../section/widget/widgets/sevenDayForecast/weatherDay/property';

let mockData: ResultData | undefined = undefined;
let mockLoading = false;

jest.mock('../section/widget/widgets/sevenDayForecast/hooks', () => ({
  useGetSevenDayForecast: () => ({
    loading: mockLoading,
    data: mockData,
  }),
}));

test('User message - Weather will update in a moment. Loading = true', () => {
  mockLoading = true;
  mockData = undefined;

  const { getByText } = render(
    <MockedProvider>
      <SevenDayForecast />
    </MockedProvider>,
  );
  expect(getByText('Weather will update in a moment.')).toBeTruthy();
});

test('User message - Weather is not available. Data = undefined', () => {
  mockData = undefined;
  mockLoading = false;

  const { getByText } = render(
    <MockedProvider>
      <SevenDayForecast />
    </MockedProvider>,
  );
  expect(getByText('Weather is not available.')).toBeTruthy();
});

test('Render SevenDayForecast', () => {
  mockData = {
    city: 'Las Vegas',
    state: 'Nevada',
    headerTemp: 87,
    headerImage: 97,
    headerTitle: 'Clear',
    days: [],
  };
  mockLoading = false;

  const { getByText } = render(
    <MockedProvider>
      <SevenDayForecast />
    </MockedProvider>,
  );
  expect(getByText('Las Vegas, Nevada')).toBeTruthy();
});

test('Render WeatherDay', () => {
  const component = renderer.create(
    <WeatherDay
      dateTime={1626271200000}
      hiTemp={100}
      humidity={30}
      image={65}
      lowTemp={88}
      precipitation={10}
      title="Sunny"
      ultraviolet="Very High"
      wind={15}
      windDirection="SSE"
    />,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('Render WeatherDay Property', () => {
  const component = renderer.create(<Property label="Precipitation" value="20%" />);
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
