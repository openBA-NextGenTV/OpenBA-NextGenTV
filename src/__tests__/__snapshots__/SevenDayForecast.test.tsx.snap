// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render WeatherDay 1`] = `
<div
  className="sc-gSYDnn jcVXrT"
>
  <div
    className="sc-laZMeE gZLYRR"
  >
    <div
      className="sc-iNiQyp jXkHnm"
    >
      Wednesday
    </div>
    <div
      className="sc-jffHpj fQbldB"
    >
      July 14
    </div>
  </div>
  <i
    className="sc-eJocfa fsxsKJ wi wi-stk-65"
  >
    <span
      className="sc-oeezt"
    />
  </i>
  <div
    className="sc-hhIiOg glQoIs"
  >
    Sunny
  </div>
  <div
    className="sc-eirqVv eIOKgU"
  >
    <span
      className="sc-lbVvki jeWfq"
    >
      100
      °
    </span>
    <span
      className="sc-gGLxEB iJublt"
    >
      88
      °
    </span>
  </div>
  <div
    className="sc-khIgEk jDxXGL"
  >
    <div
      className="sc-hTRkXV iTIxcm"
    >
      Precipitation
    </div>
    <div
      className="sc-jgPyTC boJKKo"
    >
      10%
    </div>
  </div>
  <div
    className="sc-khIgEk jDxXGL"
  >
    <div
      className="sc-hTRkXV iTIxcm"
    >
      Wind
    </div>
    <div
      className="sc-jgPyTC boJKKo"
    >
      15 mph SSE
    </div>
  </div>
  <div
    className="sc-khIgEk jDxXGL"
  >
    <div
      className="sc-hTRkXV iTIxcm"
    >
      Humidity
    </div>
    <div
      className="sc-jgPyTC boJKKo"
    >
      30%
    </div>
  </div>
  <div
    className="sc-khIgEk jDxXGL"
  >
    <div
      className="sc-hTRkXV iTIxcm"
    >
      UV
    </div>
    <div
      className="sc-jgPyTC boJKKo"
    >
      Very High
    </div>
  </div>
</div>
`;

exports[`Render WeatherDay Property 1`] = `
<div
  className="sc-khIgEk jDxXGL"
>
  <div
    className="sc-hTRkXV iTIxcm"
  >
    Precipitation
  </div>
  <div
    className="sc-jgPyTC boJKKo"
  >
    20%
  </div>
</div>
`;
